node-wasm/src/worker/channel.rs

```diff
@@ -0,0 +1,191 @@
+                }
+            })
+        };
+
+        let onmessage = Closure::new(onmessage_callback);
+        channel.set_onmessage(Some(onmessage.as_ref().unchecked_ref()));
+        channel.start();
+        Self {
+            response_channel: Mutex::new(response_rx),
+            _onmessage: onmessage,
+            channel,
+        }
+    }
+
+    pub async fn exec(&self, command: NodeCommand) -> Result<WorkerResponse, WorkerError> {
+        let mut channel = self.response_channel.lock().await;
+        self.send(command)?;
+
+        let message: WireMessage = channel
+            .recv()
+            .await
+            .ok_or(WorkerError::ResponseChannelDropped)?;
+        message.ok_or(WorkerError::EmptyWorkerResponse)?
+    }
+
+    fn send(&self, command: NodeCommand) -> Result<(), WorkerError> {
+        let command_value =
+            to_value(&command).map_err(|e| WorkerError::CouldNotSerialiseCommand(e.to_string()))?;
+        self.channel.post_message(&command_value).map_err(|e| {
+            WorkerError::CouldNotSendCommand(e.as_string().unwrap_or("UNDEFINED".to_string()))
+        })
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct ClientId(usize);
+
+impl fmt::Display for ClientId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Client({})", self.0)
+    }
+}
+
+pub(super) enum WorkerMessage {
+    NewConnection(MessagePort),
+    Command((NodeCommand, ClientId)),
+}
+
+pub(super) struct WorkerMessageServer {
+    // same onconnect callback is used throughtout entire Worker lifetime.
+    // Keep a reference to make sure it doesn't get dropped.
+    _onconnect_callback: Closure<dyn Fn(MessageEvent)>,
+
+    // keep a MessagePort for each client to send messages over, as well as callback responsible
+    // for forwarding messages back
+    clients: Vec<WorkerClientConnection>,
+
+    // sends events back to the main loop for processing
+    command_channel: mpsc::Sender<WorkerMessage>,
+}
+
+impl WorkerMessageServer {
+    pub fn new(command_channel: mpsc::Sender<WorkerMessage>) -> Self {
+        let near_tx = command_channel.clone();
+        let onconnect_callback: Closure<dyn Fn(MessageEvent)> =
+            Closure::new(move |ev: MessageEvent| {
+                let local_tx = near_tx.clone();
+                spawn_local(async move {
+                    let Ok(port) = ev.ports().at(0).dyn_into() else {
+                        error!("received connection event without MessagePort, should not happen");
+                        return;
+                    };
+
+                    if let Err(e) = local_tx.send(WorkerMessage::NewConnection(port)).await {
+                        error!("command channel inside worker closed, should not happen: {e}");
+                    }
+                })
+            });
+
+        let worker_scope = SharedWorker::worker_self();
+        worker_scope.set_onconnect(Some(onconnect_callback.as_ref().unchecked_ref()));
```

Comment:
So, worker clients are only increasing? How do we know if a worker client got disconnected?

Reply:
https://github.com/eigerco/lumina/issues/307